sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkDateTimeUtils.scala

```diff
@@ -305,21 +305,28 @@ trait SparkDateTimeUtils {
       (segment == 0 && digits >= 4 && digits <= maxDigitsYear) ||
         (segment != 0 && digits > 0 && digits <= 2)
     }
-    if (s == null || s.trimAll().numBytes() == 0) {
+    if (s == null) {
       return None
     }
+
     val segments: Array[Int] = Array[Int](1, 1, 1)
     var sign = 1
     var i = 0
     var currentSegmentValue = 0
     var currentSegmentDigits = 0
-    val bytes = s.trimAll().getBytes
+    val bytes = s.getBytes
     var j = 0
-    if (bytes(j) == '-' || bytes(j) == '+') {
+
+    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) {
+      j += 1;
+    }
+
+    if (j < bytes.length && (bytes(j) == '-' || bytes(j) == '+')) {
       sign = if (bytes(j) == '-') -1 else 1
       j += 1
     }
-    while (j < bytes.length && (i < 3 && !(bytes(j) == ' ' || bytes(j) == 'T'))) {
+    while ( j < bytes.length &&
+      (i < 3 && !(UTF8String.isWhitespaceOrISOControl(bytes(j)) || bytes(j) == 'T'))) {
```

Comment:
It would be nice to exclude the check `UTF8String.isWhitespaceOrISOControl` from the main loop. In regular cases, dates don't have any spaces, so, don't need to check every char - isn't a whitespace or not. Can we just inline `trimAll` at the beginning? like:
```scala
    var j = 0

    while (j < bytes.length && UTF8String.isWhitespaceOrISOControl(bytes(j))) j += 1
    var end = bytes.length - 1
    while (j < end && UTF8String.isWhitespaceOrISOControl(bytes(end))) end -= 1
    if (j >= end) return None

    if (bytes(j) == '-' || bytes(j) == '+') {
      sign = if (bytes(j) == '-') -1 else 1
      j += 1
    }
    while (j < end && (i < 3 && !(bytes(j) == 'T'))) {
```

Reply:
ok, I see. Let's preserve the behaviour.